client/consensus/grandpa/justification.go

```diff
@@ -27,15 +28,66 @@ var (
 //
 // This is meant to be stored in the db and passed around the network to other
 // nodes, and are used by syncing nodes to prove authority set handoffs.
-type Justification[
+type GrandpaJustification[
 	Hash constraints.Ordered,
 	N constraints.Unsigned,
 	S comparable,
 	ID AuthorityID,
-	H Header[Hash, N]] struct {
+] struct {
 	Round           uint64
 	Commit          finalityGrandpa.Commit[Hash, N, S, ID]
-	VotesAncestries []H
+	VotesAncestries []Header[Hash, N]
+}
+
+// Type used for decoding grandpa justifications (can pass in generic Header type)
+type decodeGrandpaJustification[
+	Hash constraints.Ordered,
+	N constraints.Unsigned,
+	S comparable,
+	ID AuthorityID,
+	H Header[Hash, N],
+] GrandpaJustification[Hash, N, S, ID]
+
+func DecodeGrandpaJustification[Hash constraints.Ordered,
+	N constraints.Unsigned,
+	S comparable,
+	ID AuthorityID,
+	H Header[Hash, N],
+](encodedJustification []byte) (*GrandpaJustification[Hash, N, S, ID], error) {
+	newJustificaiton := decodeGrandpaJustification[Hash, N, S, ID, H]{}
+	err := scale.Unmarshal(encodedJustification, &newJustificaiton)
+	if err != nil {
+		return nil, err
+	}
+	return newJustificaiton.GrandpaJustification(), nil
```

Comment:
I thought you were going to put this in `GrandpaJustfication.UnmarshalSCALE()`.

Reply:
I'm not really sure what you mean by this. The `decodeJustification` is just there to remove the need to interact with `decodeGrandpaJustification` from the caller since its only purpose is for decoding and I felt making the caller deal with it could lead to confusion.

I'm not sure how I would add `GrandpaJustification()` to `GrandpaJustfication.UnmarshalSCALE()` since this function has no context of an outer `GrandpaJustification` type.

I might be misunderstanding what you are suggesting here, so if I am could you elaborate more or suggest how to overcome what I said above?

Reply:
I was thinking you could just `scale.Unmarshal` into a `GrandpaJustification` which would call the custom method.  However, I don't think that'll work given we need to introduce a new generic parameter.